aprx.8
.\" APRX  v.0.11
.TH aprx 8 "2007 December 25 \- v0.11"
.LO 8
.SH NAME
.B aprx
\- A receive-only APRS iGate application,
.SH SYNOPSIS
aprx [\-d[\-d]] [\-e] [\-v] [\-l syslogfacility] [\-f /etc/aprx.conf]
.SH DESCRIPTION
.B aprx
is a
.I receive-only
APRS iGate application with minimum system support technology requirements.
.PP
.IP \(bu 3
Can listen KISS-TNCs on ordinary serial ports (real ones, or USB ones, all
work as long as they respond to normal UNIX serial port ioctl()s.)
Maximum number of separate serial ports used for radio modem communication
is 16, limit is easy to increase in source code.
.IP \(bu 3
On Linux machine with kernel internal AX.25 protocol support, does want to
listen on it with promiscuous mode, and in order to use that, must be started
as
.I root
user.
The AX.25 socket listening is
.I not
configurable, it is always on in Linux systems.
.IP \(bu 3
Does not require machine to have AX.25 protocol support internally!
(Thus this works also on e.g. BSD machines without PF\_AX25 sockets.)
.IP \(bu 3
Does not require machine to have any other software in it, than things in libc.
In particular no special AX.25 libraries at all, nor widgets or even C++ runtime.
.IP \(bu 3
Connects with one callsign-ssid pair to APRS-IS core for all received radio
ports.
.IP \(bu 3
Knows that messages with following tokens in AX.25 VIA fields are not to be
relayed into APRS-IS network:
.RS 9
.B "RFONLY, NOGATE, TCPIP, TCPXX"
.RE
.IP \(bu 3
Knows that following source address prefixes are bogus and thus messages with
them are to be junked:
.RS 9
.B "WIDE, RELAY, TRACE, TCPIP, TCPXX, NOCALL, N0CALL"
.RE
.IP \(bu 3
Drops all
.I query
messages ("?"), as well as all 3rd party messages ("}").
.IP \(bu 3
Has built-in "Erlang monitor" mechanism, that logs to syslog a interface
specific channel occupancy.   Optionally can log to STDOUT and/or to syslog.
.PP
.SH OPTIONS
The
.B aprx
has following runtime options:
.TP
.B "\-d"
Turn on verbose debugging, outputs data to STDOUT.
.TP
.B "\-dd"
the "more debug" mode shows also details of network interaction with
the APRS-IS network service.
.TP
.B "\-e"
.I "Erlang output"
prints 10 minute and 60 minute traffic accumulation byte counts, and guestimates
on channel occupancy, alias "Erlang".
These outputs are sent to STDOUT, which system operator may choose to log elsewere.
This is independent if the "\-l" option below.
.TP
.B "\-l logfacilityname"
Defines
.RB syslog (3)
facility code used by the erlang reporter by defining its name.
Default value is:
.BR NONE ,
and accepted values are:
.BR LOG_DAEMON ,
.BR LOG_FTP ,
.BR LOG_LPR ,
.BR LOG_MAIL ,
.BR LOG_NEWS ,
.BR LOG_USER ,
.BR LOG_UUCP ,
.BR LOG_LOCAL0 ,
.BR LOG_LOCAL1 ,
.BR LOG_LOCAL2 ,
.BR LOG_LOCAL3 ,
.BR LOG_LOCAL4 ,
.BR LOG_LOCAL5 ,
.BR LOG_LOCAL6 ,
.BR LOG_LOCAL7 .
That list is subject to actual facility code set in the system,
and in any case if you specify a code that is not known, then the program
will complain during the startup, and report it.
This is independent of the "\-e" option above.
.TP
.B "\-v"
Verbose logging of received traffic to STDOUT.
Lines begin with reception timestamp (UNIX time_t seconds), then TAB,
and either data as is, or with prefix byte: "*" for "discarded due to data content",
or possibly "#" for "discarded due to APRS-IS being unreachable".
.TP
.B "\-f /etc/aprx.conf"
Configuration file, given path is built-in default, and can be overridden by the program runner.
.PP
Use parameter set 
.B "\-ddv"
to test new configuration by running it synchronously to console.

.SH CONFIGURATION FILE
The configuration file is used to setup the program to do its job.
Some configurations can be done without the file, namely a channel
.I "Erlang monitor"
siphoning APRS packets off the Linux kernel AX.25 network stack.
.PP
.PP
On the configuration file, following additional notes
.IP \(bu 2
Most parameters are singletons, that is, there can be only one.
.IP \(bu 2
There can be up to (by default) 16 entries of "serialport".
.IP \(bu 2
There can be unlimited number of "netbeacon" lines.
.IP \(bu 2
There can be unlimited number of "ax25-filter" lines.
.IP \(bu 2
There can be unlimited number of "ax25-rxport" lines.
.PP
The configuration file is read upon program start, ever latter.
.nf
\fC
#
#  Sample configuration file for the APRX \-\- an Rx\-only APRS iGate
#
#

# The  mycall  parameter: 
# Station call\-id used for relaying APRS frames into APRS\-IS.
#
#mycall           N0CALL\-1

# APRS-IS server name and portnumber.
# Every reconnect does re\-resolve the name to IP address.
# There can be up to 4 system definitions without code modification,
# they are used in round\-robin fashion.  Heartbeat and filter
# definitions must follow each server definition.
#
#aprsis\-server    rotate.aprs.net  14580

# Some APRS\-IS servers tell every about 20 seconds to all contact
# ports that they are there and alive. Others are just silent.
# Enable only if the server you use does present heartbeat.
# Recommended value 3*"heartbeat" + some  \-> 120 (seconds)
#
#aprsis\-heartbeat\-timeout   120

# APRS\-IS server may support some filter commands.  Although this
# program does not transmit out to RF, filter rules can be used to
# ensure that there is sufficient dataflow from APRS\-IS server to
# this program that it very likely will not timeout within network
# monitoring timeout..
#
#aprsis\-filter "some filter specs in quotes"

# AX.25 filters block selected messages matching on selected regular
# expressions.  The expressions are case sensitive, and AX.25 address
# elements are in all uppercase text.  There can be unlimited number
# of patterns, type fields are four: "source", "destination", "via",
# and "data".  These patterns can be used in addition to built\-in
# hard\-coded reject rules listed in documentation.
#
#ax25\-filter source      "^NOCALL"
#ax25\-filter destination "^NOCALL"
#ax25\-filter via         "^NOGATE"
#ax25\-filter data        "^\\\\?"

# ax25\-rxport limits reception on listed AX.25 ports, if system
# happens to use AX.25 ports also for other purposes than APRS.
# If this option is not used, all reception ports are accepted.
# Number of port definitions here is unlimited.
#
#ax25\-rxport ax0
#ax25\-rxport ax1

# rflog defines a rotatable file into which all RF-received packets
# are logged.
#
#rflog /tmp/aprx\-rf.log

# aprxlog defines a rotatable file into which most important 
# events on APRS\-IS connection are logged, namely connects and
# disconnects.
#
#aprxlog /tmp/aprx.log

# erlangfile defines a mmap():able binary file, which stores
# running sums of interfaces upon which the channel erlang
# estimator runs, and collects data.
# Depending on the system, it may be running on a filesystem
# that actually retains data over reboots, or it may not.
# With this backing store, the system does not loose cumulating
# erlang data over the current period, if the restart is quick,
# and does not stradle any exact minute.
# (Do restarts at 15 seconds over an even minute..)
# This file is around 0.5 MB per each interface talking APRS.
# Things go BADLY WRONG if this file can not be created or
# it is corrupted!
#
# Built\-in default value is: /tmp/aprs\-erlang.dat
#
erlangfile /tmp/aprx\-erlang.dat

# erlang\-loglevel is config file edition of the "\-l" option
# pushing erlang data to syslog(3).
# Valid values are (possibly) following: NONE, LOG_DAEMON,
# LOG_FTP, LOG_LPR, LOG_MAIL, LOG_NEWS, LOG_USER, LOG_UUCP,
# LOG_LOCAL0, LOG_LOCAL1, LOG_LOCAL2, LOG_LOCAL3, LOG_LOCAL4,
# LOG_LOCAL5, LOG_LOCAL6, LOG_LOCAL7.  If the parameter value is
# not acceptable, list of accepted values are printed at startup.
#
#erlang\-loglevel NONE

# erlang\-log1min option logs to syslog/file also 1 minute
# interval data from the program. (In addition to 10m and 60m.)
#
#erlang\-log1min

# The  serialport  option.  Parameters are:
#   \- /dev/ttyUSB1    \-\- tty device
#   \- 19200           \-\- baud rate, supported ones are:
#                        1200, 2400, 4800, 9600, 19200, 38400
#   \- 8n1             \-\- 8\-bits, no parity, one stop\-bit,
#                        no other supported modes
#   \- KISS/XORSUM/BPQCRC/SMACK/CRC16  \-\- KISS mode
#
# There can be up to 16 serialport definitions in this file!
#
#serialport   /dev/ttyUSB1  19200 8n1    KISS

# Additional options for the "serialport" line.
#
# "initstring" is of two parts, the keyword, and then a string.
#    initstring "\\xC0\\xC0\\xFF\\xC0\\r\\nMO 0\\rKISS $01\\r"
#
#  "KISS"                  \- plain basic KISS mode
#  "XORSUM" alias "BPQCRC" \- KISS with BPQ "CRC" byte
#  "SMACK" alias "CRC16"   \- KISS with better CRC


# The  netbeacon  option.
# Parameter string (in quotes) is sent to network (without quotes)
# at varying intervals \-\-  1200\-1800 seconds in between restransmits.
# This interval is intentionally randomized.
#
# There can be multiple netbeacon options.
# Symbol  R&   is for "rx\-only iGate"
#
#netbeacon  "!6016.35NR02506.36E&aprx Rx\-only 'iGate'"

\fR
.fi
.PP
In the configuration file there is special treatment for quoted strings.
They are stripped of the outer quotes, and "\fC\\\fR" character is processed within
the source string to produce an output string.
The escapes are:
.TP
.B "\fC\\\\n"
Produces newline character (Control-J) on the output string.
.TP
.B "\fC\\\\r"
Produces carriage return character (Control-M) on the output string.
.TP
.B "\fC\\\\\\\\"
Places a back-slash on the output string.
.TP
.B "\fC\\\\""
Places  a double-quote on the output string.
.TP
.B "\fC\\\\'"
Places a single-quote on the output string.
.TP
.B "\fC\\\\xHH"
Lower-case "x" precedes two hex digits which ensemble is then converted to a single byte in the output string.
.PP
The complex encodings are for possible init-strings of the external devices,
.I "however: a nul byte is not possible to produce as it terminates a string!"
( = "\fC\\x00\fR" )
.PP
A configuration token without surrounding quotes does not understand the backslash escapes.


.SH NOTES: ERLANG
The
.I Erlang
is telecom measurement of channel occupancy, and in this application sense
it does tell how much traffic there is on the radio channel.
.PP
Most radio transmitters are not aware of all transmitters on channel,
and thus there can happen a collision causing loss of both messages.
The higher the channel activity, the more likely that collision is.
For further details, refer to statistical mathematics books, or perhaps
on Wikipedia.
.PP
In order to measure channel activity, the
.B aprx
program suite has these built-in statistics counter and summary estimators.
.PP
The
.I Erlag
value that the estimators present are likely somewhat
.I underestimating
the true channel occupancy simply because it calculates estimate of channel
bit transmit rate, and thus a per-minute character capacity.
It does not know true frequency of bit-stuffing events of the HDLC framing,
nor each transmitter pre- and port frame PTT times. The transmitters need to
stabilize their transmit oscillators in many cases, which may take up to
around 500 ms!
The counters are not aware of this preamble-, nor postamble-times.
.PP
The HDLC bit stuffing ratio is guessed to be 8.2 bits for each 8 bits of payload.

.SH NOTES: PROGRAM NAME
Initially this program had name
.IR aprsg-ng ,
which was same as another (less low-tech C++ approach) had.

.SH BUGS
The
.IR Erlang -monitor
mechanisms are of rudimentary quality, and can seriously underestimate the channel occupancy.

.SH SEE ALSO
Couple web sites:
.IR "http://www.aprs-is.net/" ,
.I "http://www.aprs2.net/"
.PP
.BR aprx-stat (8)

.SH AUTHOR
This little piece was written by
.I "Matti Aarnio, OH2MQK"
during a dark and rainy fall and winter of 2007-2008 after a number
of discussions grumbling about current breed of available software
for APRS iGate use in Linux (or of any UNIX) platforms.
.PP
Principal contributors and test users include:
.IR "Pentti Gronlund, OH3BK" ,
.IR "Reijo Hakala, OH1GWK" .


ChangeLog
2007-12-29  Matti Aarnio - OH2MQK - KP20NG  <mea@mea.h.zmailer.org>

	* PROTOCOLS:
	    Writeup of existing (and planned) protocols that this
	    software uses.

	* ax25.c:
	    Correct processing of 3rd-party frames.
	    Also separate TNC2 formatted frame Rx-igateing rules
	    from AX.25-to-TNC2 format translation routine.

	* ax25.c, beacon.c, aprsis.c:
	    Centralize the APRSIS communication line ending CRLF char pair
	    addition into common code, not distributed all over.

2007-12-25  Matti Aarnio - OH2MQK - KP20NG  <mea@mea.h.zmailer.org>

	* version aprx-v0.11

	* TODO, README:
	    Cleanup

	* config.c, netax25.c, aprx.8, aprx.conf:
	    Config option "ax25-rxport" - Limits acceptance of APRS
	    packets only from listed Linux AX.25 ports.

	* ttyreader.c, aprx.conf, aprx.8:
	    Reworked a bit of the serialport config options.
	    Now initstring, and various KISS modes are configurable

	* erlang.c:
	    Carefull approach on erlang-file opening, if file exists,
	    and is non-zero size, open it only if magics match.

	* config.c:
	    config_SKIPTEXT() terminates all scanned strings with NUL byte,
	    and moves to byte following it, if the termination byte was not
	    a NUL byte originally.

	* aprx-stat.c, aprx-stat.8:
	    Option -t  to show timestamps differently.

	* Makefile, man-to-html.sh:
	    Produce decent format HTML versions of the man-pages.

	* config.c, erlang.c, aprx.8, aprx.conf:
	    Added "erlang-log1min" option to control 1 minute interval
	    Erlang sampling logging behaviour.

	* Makefile:
	    Fix "make install" of man-pages

	* version aprx-v0.10

	* aprx-stat.c, aprx-stat.8, erlang.c, Makefile:
	    Statistics reporter tool.

	* aprsis.c, config.c, aprx.8, aprx.conf:
	    Multiple aprsis-server  config definitions are now
	    supported, and used in round-robin fashion.

	* ax25.c, config.c, aprx.8, aprx.conf:
	    Config option  ax25-filter

	* ttyreader.c, erlang.c:
	    Magic channel capacity constant expressions updated..

2007-12-23  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprx.c, aprx.h, aprsis.c, aprx.8, aprx.conf, ax25.c, erlang.c:
	    Multiple configuration file options, rf-log, and aprx-log -files,
	    related documentation.

	* config.c, aprx.8:
	    Special quoted-string escape processing.

	* version aprx-v0.08

	* erlang.c, aprx.c, aprx.h:
	    Erlang data has now a mmap():ed filesystem based backing-store.
	    Erlang-data can be syslog()ed, and independently of that,
	    it can be printed on STDOUT.  Default syslog facility is "NONE".
	
	* Great Rename -- the thing is now called:  aprx

2007-12-06  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprsg.8, erlang.c, aprsg.c, netax25.c, ttyreader.c:
	    Erlang logging now uses syslog(3), unless explicitely told to use
	    output to stdout.  Also the Erlang log format was altered a bit,
	    now it reports also number of packets in the interval.

	* erlang.c, Makefile:
	    "make ERLANG1=1" compiles in also 1 minute erlang logging interval.

	* aprsis.c:
	    Code refactored to put the APRS-IS communication into its own
	    fork()ed sub-process communicating via a socketpair() with the
	    main loop.   Now reconnection time with the APRS-IS server does
	    not affect functionality of the main loop.

2007-12-05  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* ChangeLog
	   Opened for the first time.  Version 0.06.



trunk/aprx.c
/* **************************************************************** *
 *                                                                  *
 *  APRX -- 2nd generation receive-only APRS-i-gate with            *
 *          minimal requirement of esoteric facilities or           *
 *          libraries of any kind beyond UNIX system libc.          *
 *                                                                  *
 * (c) Matti Aarnio - OH2MQK,  2007                                 *
 *                                                                  *
 * **************************************************************** */

#include "aprx.h"

/* Bits used only in the main program.. */
#include <signal.h>

time_t now; /* this is globally used */
int debug;
int verbout;
int erlangout;
char *rflogfile;
char *aprxlogfile;

#define CFGFILE "/etc/aprx.conf"


struct pollfd polls[MAXPOLLS];

int die_now;

const char *version = "aprx-v0.12";


static void sig_handler(int sig)
{
  die_now = 1;
}

static void usage(void)
{
	printf("aprx: [-d][-d][-e][-v][-l logfacility] [-f %s]\n", CFGFILE);
	printf("    -f %s:  where the configuration is\n", CFGFILE);
	printf("    -v:  Outputs textual format of received packets, and data on STDOUT.\n");
	printf("    -e:  Outputs raw ERLANG-report lines on SYSLOG.\n");
	printf("    -l ...: sets syslog FACILITY code for erlang reports, default: LOG_DAEMON\n");
	printf("    -d:  turn debug printout on, use to verify config file!\n");
	printf("         twice: prints also interaction with aprs-is system..\n");
	exit(64); /* EX_USAGE */
}

int fd_nonblockingmode(int fd)
{
	int __i = fcntl(fd, F_GETFL, 0);
	if (__i >= 0) {
	  /* set up non-blocking I/O */
	  __i |= O_NONBLOCK;
	  __i = fcntl(fd, F_SETFL, __i);
	}
	return __i;
}

int main(int argc, char * const argv[]) 
{
	int i;
	struct pollfd *fds;
	int nfds, nfds2;
	const char *cfgfile = "/etc/aprx.conf";
	const char *syslog_facility = "NONE";

	now = time(NULL);

	setlinebuf(stdout);
	setlinebuf(stderr);

	/* Init the poll(2) descriptor array */
	memset(polls, 0, sizeof(polls));
	for (i = 0; i < MAXPOLLS; ++i)
	  polls[i].fd = -1;

	signal(SIGTERM, sig_handler);
	signal(SIGINT, sig_handler);
	signal(SIGHUP, sig_handler);

	while ((i = getopt(argc, argv, "def:hl:v?")) != -1) {
	  switch (i) {
	  case '?': case 'h':
	    usage();
	    break;
	  case 'd':
	    ++debug;
	    break;
	  case 'e':
	    ++erlangout;
	    break;
	  case 'l':
	    syslog_facility = optarg;
	    break;
	  case 'v':
	    ++verbout;
	    break;
	  case 'f':
	    cfgfile = optarg;
	    break;
	  default:
	    break;
	  }
	}


	erlang_init(syslog_facility);
	ttyreader_init();
	netax25_init();
	aprsis_init();

	erlang_start(1);
	readconfig(cfgfile); /* TODO: real parametrized cfg file location.. */

	/* Must be after config reading ... */
	aprsis_start();

	/* The main loop */

	while (! die_now ) {

	  time_t next_timeout;
	  now = time(NULL);
	  next_timeout = now + 30;

	  fds = polls;
	  nfds = MAXPOLLS;

	  i = ttyreader_prepoll(nfds, &fds, &next_timeout);
	  nfds -= i;
	  nfds2 = i;
	  i = aprsis_prepoll(nfds, &fds, &next_timeout);
	  nfds2 += i;
	  nfds  -= i;
	  i = beacon_prepoll(nfds, &fds, &next_timeout);
	  nfds2 += i;
	  nfds  -= i;
	  i = netax25_prepoll(nfds, &fds, &next_timeout);
	  nfds2 += i;
	  nfds  -= i;
	  i = erlang_prepoll(nfds, &fds, &next_timeout);
	  nfds2 += i;
	  nfds  -= i;

	  if (next_timeout <= now)
	    next_timeout = now + 1; /* Just to be on safe side.. */

	  i = poll(polls, nfds2, (next_timeout - now) * 1000);
	  now = time(NULL);


	  i = beacon_postpoll(nfds2, polls);
	  i = ttyreader_postpoll(nfds2, polls);
	  i = netax25_postpoll(nfds2, polls);
	  i = aprsis_postpoll(nfds2, polls);
	  i = erlang_postpoll(nfds2, polls);

	}

	exit (0);
}

